File: packages/java/src/main/java/io/greyfoundry/identifold/storage/StorageAdapter.java
package io.greyfoundry.identifold.storage;

import java.util.concurrent.CompletionStage;

public interface StorageAdapter {
    CompletionStage<Boolean> reserve(ReferenceReservation request);

    CompletionStage<ReferenceMapping> resolve(String reference, String namespace);

    CompletionStage<Long> allocate(SequenceAllocationRequest request);
}

File: packages/kotlin/src/main/kotlin/io/greyfoundry/identifold/storage/StorageAdapter.kt
package io.greyfoundry.identifold.storage

data class ReferenceReservation(
    val machineId: String,
    val namespace: String,
    val reference: String,
)

data class ReferenceMapping(
    val machineId: String,
    val namespace: String,
)

data class SequenceAllocationRequest(
    val machineId: String,
    val namespace: String,
    val referencePrefix: String,
    val scope: String?,
    val width: Int,
)

interface StorageAdapter {
    suspend fun reserve(request: ReferenceReservation): Boolean

    suspend fun resolve(reference: String, namespace: String): ReferenceMapping?

    suspend fun allocate(request: SequenceAllocationRequest): Long
}
